Python/immutability.c

```diff
@@ -0,0 +1,376 @@
+
+            if(PyUnicode_Check(value)){
+                PyObject* name = value;
+                if(PyDict_Contains(globals, name)){
+                    value = PyDict_GetItem(globals, name);
+                    if(PyDict_SetItem(frozen_globals, name, value)){
+                        Py_DECREF(frozen_builtins);
+                        Py_DECREF(frozen_globals);
+                        return NULL;
+                    }
+                }
+            }
+        }
+    }
+
+    if(PyList_Append(frontier, frozen_globals)){
+        goto nomemory;
+    }
+
+    f->func_globals = frozen_globals;
+    Py_DECREF(globals);
+
+    if(PyList_Append(frontier, frozen_builtins)){
+        goto nomemory;
+    }
+
+    f->func_builtins = frozen_builtins;
+    Py_DECREF(builtins);
+
+    Py_RETURN_NONE;
+
+nomemory:
+    Py_XDECREF(frozen_builtins);
+    Py_XDECREF(frozen_globals);
+    Py_XDECREF(f_stack);
+    return PyErr_NoMemory();
+}
+
+static int freeze_visit(PyObject* obj, void* frontier)
+{
+    if(!_Py_IsImmutable(obj)){
+        if(PyList_Append((PyObject*)frontier, obj)){
+            PyErr_NoMemory();
+            return -1;
+        }
+    }
+
+    return 0;
+}
+
+PyObject* _Py_Freeze(PyObject* obj)
+{
+    if(_Py_IsImmutable(obj) && _Py_IsImmutable(Py_TYPE(obj))){
+        Py_RETURN_NONE;
+    }
+
+    PyObject* frontier = PyList_New(0);
+    if(frontier == NULL){
+        return PyErr_NoMemory();
+    }
+
+    if(PyList_Append(frontier, obj)){
+        Py_DECREF(frontier);
+        return PyErr_NoMemory();
+    }
+
+    PyObject* frozen_importlib = PyImport_ImportModule("_frozen_importlib");
+    if(frozen_importlib == NULL){
+        Py_DECREF(frontier);
+        return NULL;
+    }
+
+    PyObject* blocking_on = PyObject_GetAttrString(frozen_importlib, "_blocking_on");
+    if(blocking_on == NULL){
+        Py_DECREF(frozen_importlib);
+        Py_DECREF(frontier);
+        return NULL;
+    }
+
+    PyObject* module_locks = PyObject_GetAttrString(frozen_importlib, "_module_locks");
+    if(module_locks == NULL){
+        Py_DECREF(blocking_on);
+        Py_DECREF(frozen_importlib);
+        Py_DECREF(frontier);
+        return NULL;
+    }
+
+    Py_DECREF(frozen_importlib);
+
+    while(PyList_Size(frontier) != 0){
+        PyObject* item = pop(frontier); // item.rc = x + 1
+
+        if(item == blocking_on ||
+           item == module_locks){
+            // the module lock and blocking on dictionaries must remain mutable or else
+            // we will not be able to import modules
+            continue;
+        }
+
+        PyTypeObject* type = Py_TYPE(item);
+        traverseproc traverse;
+        PyObject* type_op = NULL;
+
+        if(_Py_IsImmutable(item)){
+            goto handle_type;
+        }
+
+        _Py_SetImmutable(item);
+
+
+        if(is_c_wrapper(item)) {
+            // C functions are not mutable, so we can skip them.
+            continue;
+        }
```

Comment:
Could we mark all c-wrapper types as immutable by default? I remember that we talked about this some time ago and that this was planned for "Phase 2" before we changed the order and approach.

Reply:
Yeah I should look at this again too. It may be that we can just mark them immutable in the normal way and call it a day instead of leaving them around and requiring this check.